runtime/drivers/druid/druidsqldriver/druid_api_sql_driver.go

```diff
@@ -40,193 +55,285 @@ type sqlConnection struct {
+	}
+}
+
+func (c *sqlConnection) QueryContext(ctx context.Context, query string, args []driver.NamedValue) (driver.Rows, error) {
+	// total sum is 126 seconds (sum(2*2^x) from 0 to 5 inclusive)
+	re := retrier.NewRetrier(6, 2*time.Second, &coordinatorHTTPCheck{
+		c: c,
+	})
+	return re.RunCtx(ctx, func(ctx2 context.Context) (driver.Rows, retrier.Action, error) {
+		dr := newDruidRequest(query, args)
+		b, err := json.Marshal(dr)
 		if err != nil {
+			return nil, retrier.Fail, err
+		}
+
+		bodyReader := bytes.NewReader(b)
+
+		context.AfterFunc(ctx, func() {
+			tctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+			defer cancel()
+			r, err := http.NewRequestWithContext(tctx, http.MethodDelete, c.dsn+"/"+dr.Context.SQLQueryID, http.NoBody)
+			if err != nil {
+				return
+			}
+
+			resp, err := c.client.Do(r)
+			if err != nil {
+				return
+			}
 			resp.Body.Close()
-			return nil, err
+		})
+
+		req, err := http.NewRequestWithContext(ctx, http.MethodPost, c.dsn, bodyReader)
+		if err != nil {
+			return nil, retrier.Fail, err
 		}
 
-		types := toStringArray(obj.([]any))
-
-		transformers := make([]func(any) (any, error), len(columns))
-		for i, c := range types {
-			transformers[i] = identityTransformer
-			switch c {
-			case "TINYINT":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case float64:
-						return int8(v), nil
-					default:
-						return v, nil
-					}
-				}
-			case "SMALLINT":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case float64:
-						return int16(v), nil
-					default:
-						return v, nil
-					}
-				}
-			case "INTEGER":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case float64:
-						return int32(v), nil
-					default:
-						return v, nil
-					}
-				}
-			case "BIGINT":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case float64:
-						return int64(v), nil
-					default:
-						return v, nil
-					}
-				}
-			case "FLOAT":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case float64:
-						return float32(v), nil
-					case string:
-						return strconv.ParseFloat(v, 32)
-					default:
-						return v, nil
-					}
-				}
-			case "DOUBLE":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case string:
-						return strconv.ParseFloat(v, 64)
-					default:
-						return v, nil
-					}
-				}
-			case "REAL":
-				transformers[i] = func(v any) (any, error) {
-					switch v := v.(type) {
-					case string:
-						return strconv.ParseFloat(v, 64)
-					default:
-						return v, nil
-					}
+		req.Header.Add("Content-Type", "application/json")
+		resp, err := c.client.Do(req)
+		if err != nil {
+			// nolint:errorlint // there's no wrapping
+			if v, ok := err.(*url.Error); ok {
+				if errTooManyRedirects.MatchString(v.Error()) {
+					resp.Body.Close()
```

Comment:
This may panic – accessing the body when the request failed. 

It's a problem multiple places, I would serialize consider if this can be refactored to use only `defer`.

Reply:
Thanks. But `defer` won't work - see next comment.

Reply:
Okay, got it. In that case, it's worth considering if things can be re-organized in such a way that `resp.Body.Close` needs to be manually called in as few cases as possible

Reply:
Let's think about it in another PR.